### PR TITLE
When read PoolSubpage's variant fields, it should lock on PoolSubpage's head lock, instead of PoolArena's lock

### PR DESCRIPTION
**Pull request by laosijikaichele**
Motivation:

When we read `PoolSubpage`'s variant fields:`doNotDestroy` and `numAvail`, we should lock on `PoolSubpage`'s head subpage, instead of `PoolArena`'s lock. 

Because the writing on `doNotDestroy` and `numAvail` is always under `PoolSubpage`'s head subpage lock.

Modification:

1. Introduced a `int headIndex` field in `PoolSubpage` to locate the head subpage, lock on `PoolSubpage`'s head subpage instead of `PoolArena`'s lock. 
2. This also help to locate head subpage when doing `bytebuf` release. Fixed [13625](https://github.com/netty/netty/issues/13625).

Result:

See above.

### CODE REVIEW

File: buffer/src/test/java/io/netty/buffer/AlignedPooledByteBufAllocatorTest.java [hunk 1]
```diff
@@ -63,4 +66,62 @@ public void testCorrectElementSize() {
         assertTrue(a.release());
         assertTrue(b.release());
     }
+    @Test
+    public void testDirectSubpageReleaseLock() throws Exception {
```
Comment (normanmaurer) [inline buffer/src/test/java/io/netty/buffer/AlignedPooledByteBufAllocatorTest.java:70]: Could we change our visibility of things to package-private so we dont need reflection for the test ?

File: buffer/src/main/java/io/netty/buffer/PooledByteBufAllocator.java [hunk 2]
```diff
@@ -349,7 +348,7 @@ private static int validateAndCalculatePageShifts(int pageSize, int alignment) {
         }
 
         // Logarithm base 2. At this point we know that pageSize is a power of two.
-        return Integer.SIZE - 1 - Integer.numberOfLeadingZeros(pageSize);
+        return (byte) (Integer.SIZE - 1 - Integer.numberOfLeadingZeros(pageSize));
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/PooledByteBufAllocator.java:351]: The changes in this class do not look related... please revert

File: buffer/src/main/java/io/netty/buffer/PoolChunk.java [hunk 3]
```diff
@@ -610,7 +608,7 @@ void initBufWithSubpage(PooledByteBuf<T> buf, ByteBuffer nioBuffer, long handle,
         int bitmapIdx = bitmapIdx(handle);
 
         PoolSubpage<T> s = subpages[runOffset];
-        assert s.doNotDestroy;
+        assert s.doNotDestroy();
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/PoolChunk.java:611]: Imho it's very confusing that we have a method and also a field that is called `doNotDestroy`.

File: buffer/src/main/java/io/netty/buffer/PoolChunk.java [hunk 4]
```diff
@@ -684,7 +682,7 @@ static int runOffset(long handle) {
         return (int) (handle >> RUN_OFFSET_SHIFT);
     }
 
-    static int runSize(int pageShifts, long handle) {
+    static int runSize(byte pageShifts, long handle) {
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/PoolChunk.java:685]: please refer this change... not related

File: buffer/src/main/java/io/netty/buffer/PoolArena.java [hunk 5]
```diff
@@ -691,7 +691,7 @@ boolean isDirect() {
 
         @Override
         protected PoolChunk<ByteBuffer> newChunk(int pageSize, int maxPageIdx,
-            int pageShifts, int chunkSize) {
+            byte pageShifts, int chunkSize) {
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/PoolArena.java:694]: not related... please revert

File: buffer/src/main/java/io/netty/buffer/PoolArena.java [hunk 6]
```diff
@@ -629,7 +629,7 @@ private void destroyPoolChunkLists(PoolChunkList<T>... chunkLists) {
 
     static final class HeapArena extends PoolArena<byte[]> {
 
-        HeapArena(PooledByteBufAllocator parent, int pageSize, int pageShifts,
+        HeapArena(PooledByteBufAllocator parent, int pageSize, byte pageShifts,
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/PoolArena.java:632]: not related... please revert

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: buffer/src/main/java/io/netty/buffer/PoolChunk.java (hunk 3 above)
Comment (laosijikaichele) [reply buffer/src/main/java/io/netty/buffer/PoolChunk.java:611]: em... What about `isDoNotDestroy()`?

File: buffer/src/test/java/io/netty/buffer/AlignedPooledByteBufAllocatorTest.java (hunk 1 above)
Comment (laosijikaichele) [reply buffer/src/test/java/io/netty/buffer/AlignedPooledByteBufAllocatorTest.java:70]: Yes,  this can be done by changing the field `PoolArena.smallSubpagePools` to package-private.

File: buffer/src/test/java/io/netty/buffer/AlignedPooledByteBufAllocatorTest.java [hunk 7]
```diff
@@ -63,4 +66,58 @@ public void testCorrectElementSize() {
...
+                byteBuf.release();
+            }
+        });
+        t1.setDaemon(true);
+        // Intentionally lock the head subpage in main thread.
+        head.lock();
+        try {
+            t1.start();
+            // Wait max 3 seconds.
+            t1.join(1000 * 3);
```
Comment (chrisvest) [inline buffer/src/test/java/io/netty/buffer/AlignedPooledByteBufAllocatorTest.java:110]: Instead of having a mandatory 3-second wait here, we could make `PoolSubpage.lock` package-private and spin on a `while (!head.lock.hasQueuedThread(t1)) { ... }` condition, until some deadline.

File: buffer/src/test/java/io/netty/buffer/AlignedPooledByteBufAllocatorTest.java [hunk 8]
```diff
@@ -63,4 +66,58 @@ public void testCorrectElementSize() {
...
+            assertTrue(t1.isAlive(), "The t1 thread should still alive and wait for the head lock.");
+        } finally {
+            head.unlock();
+        }
+    }
+    private static PooledByteBuf<?> pooledByteBuf(ByteBuf buffer) {
```
Comment (chrisvest) [inline buffer/src/test/java/io/netty/buffer/AlignedPooledByteBufAllocatorTest.java:116]: Please put a blank line between method definitions.

**Review by chrisvest — CHANGES_REQUESTED** [review]
I had some comments. Please take a look.

File: buffer/src/test/java/io/netty/buffer/AlignedPooledByteBufAllocatorTest.java [hunk 9]
```diff
@@ -63,4 +66,58 @@ public void testCorrectElementSize() {
         assertTrue(a.release());
         assertTrue(b.release());
     }
+    @Test
+    public void testDirectSubpageReleaseLock() throws Exception {
```
[the remark this answers is not in the document]
Comment (laosijikaichele) [reply buffer/src/test/java/io/netty/buffer/AlignedPooledByteBufAllocatorTest.java:70]: I used to add an assertion in method `PoolChunk.free()`: `assert before_PR_subpge_head == after_PR_subpge_head`, and the test `AlignedPooledByteBufAllocatorTest.testCorrectElementSize()` failed the assertion. So I think it's related to this `AlignedPooledByteBufAllocatorTest` class, and added the new test case here too.

File: buffer/src/test/java/io/netty/buffer/AlignedPooledByteBufAllocatorTest.java (hunk 7 above)
Comment (laosijikaichele) [reply buffer/src/test/java/io/netty/buffer/AlignedPooledByteBufAllocatorTest.java:110]: But imho, we still need to specified a mandatory wait 'time' or 'times' as the deadline signal, to decide when to end the `while (!head.lock.hasQueuedThread(t1)) { ... }` looping.

File: buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java [hunk 10]
```diff
@@ -39,6 +39,7 @@
 
 import static io.netty.buffer.PoolChunk.runOffset;
 import static io.netty.buffer.PoolChunk.runPages;
+import static io.netty.buffer.PoolChunk.RUN_OFFSET_SHIFT;
```
[the remark this answers is not in the document]
Comment (laosijikaichele) [reply buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java]: Done.

File: buffer/src/test/java/io/netty/buffer/AlignedPooledByteBufAllocatorTest.java (hunk 8 above)
Comment (laosijikaichele) [reply buffer/src/test/java/io/netty/buffer/AlignedPooledByteBufAllocatorTest.java:117]: Done.

File: buffer/src/test/java/io/netty/buffer/AlignedPooledByteBufAllocatorTest.java (hunk 7 above)
Comment (laosijikaichele) [reply buffer/src/test/java/io/netty/buffer/AlignedPooledByteBufAllocatorTest.java:110]: But imho, even though, we still need a mandatory wait 'time' or 'times' to decide when to end the `while (!head.lock.hasQueuedThread(t1)) { ... }` looping.

File: buffer/src/test/java/io/netty/buffer/AlignedPooledByteBufAllocatorTest.java (hunk 1 above)
Comment (laosijikaichele) [reply buffer/src/test/java/io/netty/buffer/AlignedPooledByteBufAllocatorTest.java:70]: Done.

File: buffer/src/main/java/io/netty/buffer/PoolChunk.java (hunk 4 above)
Comment (laosijikaichele) [reply buffer/src/main/java/io/netty/buffer/PoolChunk.java]: Done.

File: buffer/src/main/java/io/netty/buffer/PoolArena.java (hunk 5 above)
Comment (laosijikaichele) [reply buffer/src/main/java/io/netty/buffer/PoolArena.java]: Done.

File: buffer/src/main/java/io/netty/buffer/PoolArena.java (hunk 6 above)
Comment (laosijikaichele) [reply buffer/src/main/java/io/netty/buffer/PoolArena.java]: Done.

File: buffer/src/test/java/io/netty/buffer/AlignedPooledByteBufAllocatorTest.java (hunk 7 above)
Comment (chrisvest) [reply buffer/src/test/java/io/netty/buffer/AlignedPooledByteBufAllocatorTest.java:108]: The code can set itself a deadline and check it with `System.nanoTime()`. I think that should suffice to avoid infinite looping.

Comment (laosijikaichele) [reply buffer/src/test/java/io/netty/buffer/AlignedPooledByteBufAllocatorTest.java]: @chrisvest , done, PTAL.

File: buffer/src/test/java/io/netty/buffer/AlignedPooledByteBufAllocatorTest.java [hunk 11]
```diff
@@ -63,4 +66,64 @@ public void testCorrectElementSize() {
         assertTrue(a.release());
         assertTrue(b.release());
     }
+    @Test
```
Comment (normanmaurer) [inline buffer/src/test/java/io/netty/buffer/AlignedPooledByteBufAllocatorTest.java:69]:
```suggestion

    @Test
```

**Review by normanmaurer — CHANGES_REQUESTED** [review]

**Review by normanmaurer — APPROVED** [review]

**Review by chrisvest — APPROVED** [review]

**Comment by chrisvest** [comment]
@laosijikaichele Thanks!

**Comment by echooymxq** [comment]
@laosijikaichele @chrisvest 
this PR is also work for https://github.com/netty/netty/issues/13777?

**Comment by laosijikaichele** [comment]
@echooymxq I am not sure, but it might be related, in some cases, it locks on wrong `head` when releasing direct buffer, which will make the modification of `numAvail` not thread safe, please refer https://github.com/netty/netty/issues/13625.
